### PR TITLE
Dump trees

### PR DESCRIPTION
**Pull request by ncordon**
Code for writing and reading taxonomy trees to files.

Also this PR adds generation of trees to the mirroring of files, and a better ordering of tests, so that mirroring is performed before checking taxonomy files are OK.

I will add documentation in a separate PR.

Right now it is expected that tests do not compile since we need an in-cascade releasing process for all descendants of `files` package: `files`, `forests` and`s3`

### CODE REVIEW

File: deps.sbt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 libraryDependencies ++= Seq(
-  "ohnosequences" %% "s3"      % "0.1.0-4-gaa153b0",
-  "ohnosequences" %% "forests" % "0.1.0"
+  "ohnosequences" %% "s3"      % "0.1.0-11-ge66487e-SNAPSHOT",
+  "ohnosequences" %% "forests" % "0.1.0-2-gdbfdd11-SNAPSHOT"
```
Comment (agarciamontoro) [inline deps.sbt:3]: Should we release new versions of these packages?

File: src/main/scala/io.scala [hunk 2]
```diff
@@ -151,4 +170,75 @@ case object io {
...
+      shapeResult.map { shapeFile =>
+        (dataFile, shapeFile)
+      }
+    }
+  }
+
+  def readTaxTreeFromFiles(
+      dataFile: File,
+      shapeFile: File
+  ): FileError + (SerializationError + TaxTree) = {
```
Comment (agarciamontoro) [inline src/main/scala/io.scala]: I like to group the errors in the left side, as in `(FileError + SerializationError) + TaxTree`, just to be able to directly map the either and work with the result. Although this is just a preference, we should agree to always implement these values in one or the other way. Just saying it here to open a discussion :) What do you prefer?

File: src/test/scala/checkOldVersions.scala [hunk 3]
```diff
@@ -1,23 +1,41 @@
...
+      case (version, namesObj, nodesObj) =>
+        assert(
+          objectExists(namesObj) && objectExists(nodesObj),
+          s"S3 objects names.dmp and nodes.dmp do not exist for version ${version.name}"
+        )
+    }
+  }
+
+  test(
+    "data.tree and shape.tree exist in S3 for all versions. If not, create them") {
```
Comment (agarciamontoro) [inline src/test/scala/checkOldVersions.scala]: This does not create anything, isn't it?

File: src/test/scala/mirrorAndCheck.scala [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+import org.scalatest.{CancelAfterFailure, Suites}
+
+class MirrorAndCheck
+    extends Suites(
+      // Ensure mirroring before checking files
+      new Mirror,
+      new CheckOldVersions,
+      new CheckFullTaxonomy
+    )
```
Comment (agarciamontoro) [inline src/test/scala/mirrorAndCheck.scala:11]: Ah, ok, this and the `DoNotDiscover` annotations make sure the order is right, have I understood it right?

**Review by agarciamontoro — APPROVED** [review]
LGTM!

I, however, would want to discuss in the near future (with @eparejatobes) this philosophy of releasing data by executing tests. We are now even relying on the order of the test suite (I see no other way of doing it right now), and executing several unit tests in a specific order is not that different from executing a single big test that does everything :(

File: deps.sbt (hunk 1 above)
Comment (ncordon) [reply deps.sbt:3]: Yep, we should

File: src/test/scala/checkOldVersions.scala (hunk 3 above)
Comment (ncordon) [reply src/test/scala/checkOldVersions.scala]: Correct! I finally left all that work to the `mirror` test, which is ordered to be executed before this check

File: src/test/scala/mirrorAndCheck.scala (hunk 4 above)
Comment (ncordon) [reply src/test/scala/mirrorAndCheck.scala:11]: The `Suites` structure makes it possible to order the tests, provided that parallel execution of tests is not run (we have it disabled in our default configuration, so yay :tada: ). The `DoNotDiscover` tells `scalatest` to not execute a suite that is marked with that annotation (if we did not include the `DoNotDiscover`, `Mirror` would be executed one time inside the `Suites` structure and another time as a normal suite, as the rests of tests do).

Comment (agarciamontoro) [reply src/test/scala/mirrorAndCheck.scala:11]: Ok, thanks!

File: src/main/scala/io.scala (hunk 2 above)
Comment (ncordon) [reply src/main/scala/io.scala]: You are right, your option is better (let's do it your way henceforth). I will update it before merging the PR

Comment (agarciamontoro) [reply src/main/scala/io.scala]: Oh, that was an easy discussion!

**Pull request merged**